libraries/AP_HAL/DSP.h

```diff
@@ -36,48 +55,54 @@ class AP_HAL::DSP {
         const uint16_t _bin_count;
         // size of the FFT window
         const uint16_t _window_size;
+        // number of harmonics of interest
+        const uint8_t _harmonics;
         // FFT data
         float* _freq_bins;
+        // derivative real data scratch space
+        float* _derivative_freq_bins;
         // intermediate real FFT data
         float* _rfft_data;
-        // estimate of FFT peak frequency
-        float _max_bin_freq;
-        // bin with maximum energy
-        uint16_t _max_energy_bin;
-        // width of the max energy peak
-        float _max_noise_width_hz;
-        // estimate of FFT second peak frequency
-        float _second_bin_freq;
-        // bin with second-most energy
-        uint16_t _second_energy_bin;
-        // width of the second energy peak
-        float _second_noise_width_hz;
+        // three highest peaks
+        FrequencyPeakData _peak_data[MAX_TRACKED_PEAKS];
         // Hanning window for incoming samples, see https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window
         float* _hanning_window;
         // Use in calculating the PS of the signal [Heinz] equations (20) & (21)
         float _window_scale;
 
         virtual ~FFTWindowState();
-        FFTWindowState(uint16_t window_size, uint16_t sample_rate);
+        FFTWindowState(uint16_t window_size, uint16_t sample_rate, uint8_t harmonics);
     };
     // initialise an FFT instance
-    virtual FFTWindowState* fft_init(uint16_t window_size, uint16_t sample_rate) = 0;
+    virtual FFTWindowState* fft_init(uint16_t window_size, uint16_t sample_rate, uint8_t harmonics) = 0;
     // start an FFT analysis
     virtual void fft_start(FFTWindowState* state, const float* samples, uint16_t buffer_index, uint16_t buffer_size) = 0;
+    // start an FFT analysis with an ObjectBuffer
+    virtual void fft_start(FFTWindowState* state, FloatBuffer& samples, uint16_t advance) = 0;
```

Comment:
why do we need to keep both variants of fft_start ? I ran across this when looking at the constructor/destructors we end up with in the chibios build. We end up with 6 copies of the FFTWindowState destructor (3 in AP_HAL, 3 in ChibiOS:DSP)

Reply:
Removed the float version